sound/soc/sof/ipc4-loader.c

```diff
@@ -160,13 +164,113 @@ static size_t sof_ipc4_fw_parse_basefw_ext_man(struct snd_sof_dev *sdev)
 	return payload_offset;
 }
 
+static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
+					 unsigned long lib_id, const guid_t *uuid)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct sof_ipc4_fw_library *fw_lib;
+	const char *fw_filename;
+	size_t payload_offset;
+	int ret, i, err;
+
+	if (!sdev->pdata->fw_lib_prefix) {
+		dev_err(sdev->dev,
+			"Library loading is not supported due to not set library path\n");
+		return -EINVAL;
+	}
+
+	if (!ipc4_data->load_library) {
+		dev_err(sdev->dev, "Library loading is not supported on this platform\n");
+		return -EOPNOTSUPP;
+	}
+
+	fw_lib = devm_kzalloc(sdev->dev, sizeof(*fw_lib), GFP_KERNEL);
+	if (!fw_lib)
+		return -ENOMEM;
+
+	fw_filename = kasprintf(GFP_KERNEL, "%s/%pUL.bin",
+				sdev->pdata->fw_lib_prefix, uuid);
+	if (!fw_filename) {
+		ret = -ENOMEM;
+		goto free_fw_lib;
+	}
+
+	ret = request_firmware(&fw_lib->sof_fw.fw, fw_filename, sdev->dev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "Library file '%s' is missing\n", fw_filename);
+		goto free_filename;
+	} else {
+		dev_dbg(sdev->dev, "Library file '%s' loaded\n", fw_filename);
+	}
+
+	payload_offset = sof_ipc4_fw_parse_ext_man(sdev, fw_lib);
+	if (payload_offset <= 0) {
+		if (!payload_offset)
+			ret = -EINVAL;
+		else
+			ret = payload_offset;
+
+		goto release;
+	}
+
+	fw_lib->sof_fw.payload_offset = payload_offset;
+	fw_lib->id = lib_id;
```

Comment:
@ujfalusi i think there's a library offset of 0x1000 for external libraries @aiChaoSONG correct me if I am wrong please

Reply:
the offset is not for library id, it is for module id in the loadable library.
The module id in a loadable lib is calculated like this: `*module_id = (library_idx << 12) + local_mod_id;`

Reply:
@ujfalusi last question, are we handling this correctly in this PR?

Reply:
> @ujfalusi last question, are we handling this correctly in this PR?

The library id is a running integer number.
The module's ID is a composite of library ID (which the module belongs to) and a running integer for the module ID.

To not mess around with the generic ext manifest parsing where we assign IDs to the modules within the library I have opted to update the module ID here at line 221.

Reply:
ok that makes me wonder line 221 is the < correct?

Reply:
@ranj063, you have good eye ;) It should have been `<<` as shift..